frame/pallet-xvm/src/evm.rs

```diff
@@ -1,61 +1,67 @@
+/// using Solidity ABI codec (https://docs.soliditylang.org/en/v0.8.16/abi-spec.html).
+pub struct EVM<I, T>(sp_std::marker::PhantomData<(I, T)>);
 
-impl<VmId, I, T, C> SyncVM<VmId, T::AccountId> for EVM<I, T, C>
+impl<I, T> SyncVM<T::AccountId> for EVM<I, T>
 where
     I: Get<VmId>,
     T: pallet_evm::Config + frame_system::Config,
-    C: XvmCodec,
 {
     fn id() -> VmId {
         I::get()
     }
 
-    fn xvm_call(
-        _context: XvmContext<VmId>,
-        from: T::AccountId,
-        to: Vec<u8>,
-        input: Vec<u8>,
-        metadata: Vec<u8>,
-    ) -> Result<Vec<u8>, Vec<u8>> {
+    fn xvm_call(context: XvmContext, from: T::AccountId, to: Vec<u8>, input: Vec<u8>) -> XvmResult {
         log::trace!(
             target: "xvm::EVM::xvm_call",
-            "Start EVM XVM: {:?}, {:?}, {:?}, {:?}",
-            from, to, input, metadata,
+            "Start EVM XVM: {:?}, {:?}, {:?}",
+            from, to, input,
         );
-        let data = C::convert(input, metadata)?;
         let value = U256::from(0u64);
         let max_fee_per_gas = U256::from(3450898690u64);
-        let gas_limit = 4000000u64;
-        let _nonce = frame_system::Pallet::<T>::account(from.clone()).nonce;
-        let evm_to: H160 = Decode::decode(&mut to.as_ref())
-            .map_err(|_| b"`to` argument decode failure".to_vec())?;
+        let gas_limit = T::GasWeightMapping::weight_to_gas(context.max_weight);
+        log::trace!(
+            target: "xvm::EVM::xvm_call",
+            "EVM xvm call gas limit: {:?} or as weight: {:?}", gas_limit, context.max_weight);
+        let evm_to = Decode::decode(&mut to.as_ref()).map_err(|_| XvmCallError {
+            error: XvmError::EncodingFailure,
+            consumed_weight: PLACEHOLDER_WEIGHT,
+        })?;
 
         let res = pallet_evm::Pallet::<T>::call(
             frame_support::dispatch::RawOrigin::Root.into(),
             H160::from_slice(&from.encode()[0..20]),
             evm_to,
-            data,
+            input,
             value,
             gas_limit,
             max_fee_per_gas,
             None,
             None,
             Vec::new(),
-        );
+        )
+        .map_err(|_e| XvmCallError {
+            error: XvmError::ExecutionError(Vec::default()), // TODO: make error mapping make more sense
+            consumed_weight: 42u64, // TODO: res.actual_weight.map(|x| x.ref_time()).unwrap_or(context.max_weight),
+        })?;
 
         log::trace!(
             target: "xvm::EVM::xvm_call",
             "EVM XVM call result: {:?}", res
         );
 
-        // TODO: return error if call failure
-        // TODO: return value in case of constant / view call
-        Ok(Default::default())
+        Ok(XvmCallOk {
+            output: Default::default(), // TODO: Fill output vec with response from the call
+            consumed_weight: 42u64, // TODO: res.actual_weight.map(|x| x.ref_time()).unwrap_or(context.max_weight),
```

Comment:
You should extract correct `consumed_weight` from the `res`.

Reply:
Yes, but next time

Reply:
Why next time?
Weights should be handled properly in v2, right?

And it's a very minor fix. I guess same question as above, but why is the "correct" code commented out?

Reply:
Unfortunately this code no more works correctly, it assumes that weight is a number, but it is struct now.
I just want merge this PR because it takes so long time and proper weighting isn’t the main aim of this PR.

Reply:
I see - however, the struct that was introduced was simply a wrapper around the old u64 value.
That was the initial step, the next one being adding the PoV size.

For this particular case, in the PR, all you need to do is this:

```
consumed_weight: res.actual_weight.map(|x| x.ref_time()).unwrap_or(context.max_weight.ref_time())
```
Just add `.ref_time()` to the `max_weight`.

Reply:
Not that simple but fixed